Public/Src/Cache/ContentStore/Library/Utils/BandwidthChecker.cs

```diff
@@ -45,93 +45,78 @@ public async Task<CopyFileResult> CheckBandwidthAtIntervalAsync(OperationContext
         {
             if (_historicalBandwidthLimitSource != null)
             {
-                var startPosition = destinationStream.Position;
+                var startPosition = tryGetPosition(destinationStream, out var pos) ? pos : 0;
                 var timer = Stopwatch.StartNew();
+                var (result, lastPosition) = await impl();
+                timer.Stop();
 
-                try
-                {
-                    return await impl();
-                }
-                finally
-                {
-                    timer.Stop();
-                    var endPosition = destinationStream.Position;
+                // Bandwidth checker expects speed in MiB/s, so convert it.
+                var bytesCopied = lastPosition - startPosition;
+                var speed = bytesCopied / timer.Elapsed.TotalSeconds / (1024 * 1024);
+                _historicalBandwidthLimitSource.AddBandwidthRecord(speed);
 
-                    // Bandwidth checker expects speed in MiB/s, so convert it.
-                    var bytesCopied = endPosition - startPosition;
-                    var speed = bytesCopied / timer.Elapsed.TotalSeconds / (1024 * 1024);
-                    _historicalBandwidthLimitSource.AddBandwidthRecord(speed);
-                }
+                return result;
             }
             else
             {
-                return await impl();
+                return (await impl()).result;
             }
 
-            async Task<CopyFileResult> impl()
+            async Task<(CopyFileResult result, long lastPosition)> impl()
             {
                 // This method should not fail with exceptions because the resulting task may be left unobserved causing an application to crash
                 // (given that the app is configured to fail on unobserved task exceptions).
                 var minimumSpeedInMbPerSec = _bandwidthLimitSource.GetMinimumSpeedInMbPerSec() * _config.BandwidthLimitMultiplier;
                 minimumSpeedInMbPerSec = Math.Min(minimumSpeedInMbPerSec, _config.MaxBandwidthLimit);
 
-                long previousPosition = 0;
+                long firstPosition = tryGetPosition(destinationStream, out var pos) ? pos : 0;

```

Comment:
Does it make sense to obtain the start position in the enclosing method not inside the if, but on top of it and the impl can just reference it?